core/src/main/scala/kafka/log/Log.scala

```diff
@@ -445,17 +448,20 @@ class Log(@volatile var dir: File,
         producerStateManager.takeSnapshot()
       }
     } else {
-      val currentTimeMs = time.milliseconds
-      producerStateManager.truncateAndReload(logStartOffset, lastOffset, currentTimeMs)
+      val isEmptyBeforeTruncation = producerStateManager.isEmpty && producerStateManager.mapEndOffset >= lastOffset
```

Comment:
That's only the case when doing the initial loading. Truncation after initial load would hit this case.